File: lib/code_puppy_control/config/agents.ex
defmodule CodePuppyControl.Config.Agents do
  @moduledoc """
  Agent configuration accessors.

  Manages the default agent, agent directories, and related preferences.

  ## Config keys in `puppy.cfg`

  - `default_agent` — name of the default agent (default `"code-puppy"`)
  - `puppy_name` — display name for the puppy
  - `owner_name` — display name for the owner
  """

  alias CodePuppyControl.Config.{Loader, Paths}

  @default_agent_name "code-puppy"

  # ── Default agent ───────────────────────────────────────────────────────

  @doc """
  Return the default agent name. Falls back to `"code-puppy"`.
  """
  @spec default_agent() :: String.t()
  def default_agent do
    Loader.get_value("default_agent") || @default_agent_name
  end

  @doc """
  Set the default agent name.
  """
  @spec set_default_agent(String.t()) :: :ok
  def set_default_agent(agent_name) when is_binary(agent_name) do
    CodePuppyControl.Config.Writer.set_value("default_agent", agent_name)
  end

  # ── Personalization ─────────────────────────────────────────────────────

  @doc """
  Return the puppy's display name. Defaults to `"Puppy"`.
  """
  @spec puppy_name() :: String.t()
  def puppy_name do
    Loader.get_value("puppy_name") || "Puppy"
  end

  @doc """
  Return the owner's display name. Defaults to `"Master"`.
  """
  @spec owner_name() :: String.t()
  def owner_name do
    Loader.get_value("owner_name") || "Master"
  end

  # ── Agent directories ───────────────────────────────────────────────────

  @doc """
  Return the user-level agents directory. Ensures it exists.
  """
  @spec user_agents_dir() :: String.t()
  def user_agents_dir do
    dir = Paths.agents_dir()
    File.mkdir_p!(dir)
    dir
  end

  @doc """
  Return the project-local agents directory if `.code_puppy/agents/` exists
  in CWD, or `nil`.
  """
  @spec project_agents_dir() :: String.t() | nil
  def project_agents_dir do
    Paths.project_agents_dir()
  end

  @doc """
  Return all agent directories to search (project first, then user).
  """
  @spec agent_search_paths() :: [String.t()]
  def agent_search_paths do
    [project_agents_dir(), user_agents_dir()]
    |> Enum.filter(&(&1 != nil))
  end
end

File: code_puppy/config/agents.py
"""Agent configuration accessors.

Mirrors ``CodePuppyControl.Config.Agents`` in the Elixir runtime.

Manages the default agent, agent directories, and personalization settings.

Config keys in puppy.cfg:

- ``default_agent`` — name of the default agent (default ``"code-puppy"``)
- ``puppy_name`` — display name for the puppy
- ``owner_name`` — display name for the owner
"""

from __future__ import annotations

import os

from code_puppy.config.loader import (
    _registered_cache,
    get_value,
    set_config_value,
)
from code_puppy.config.paths import (
    _path_agents_dir,
)

__all__ = [
    "get_default_agent",
    "set_default_agent",
    "get_puppy_name",
    "get_owner_name",
    "get_user_agents_directory",
    "get_project_agents_directory",
    "get_puppy_token",
    "set_puppy_token",
]


# ---------------------------------------------------------------------------
# Default agent
# ---------------------------------------------------------------------------


def get_default_agent() -> str:
    """Return the default agent name. Falls back to ``"code-puppy"``."""
    return get_value("default_agent") or "code-puppy"


def set_default_agent(agent_name: str) -> None:
    """Set the default agent name in puppy.cfg."""
    set_config_value("default_agent", agent_name)


# ---------------------------------------------------------------------------
# Personalization
# ---------------------------------------------------------------------------


def get_puppy_name() -> str:
    """Return the puppy's display name. Defaults to ``"Puppy"``."""
    return get_value("puppy_name") or "Puppy"


def get_owner_name() -> str:
    """Return the owner's display name. Defaults to ``"Master"``."""
    return get_value("owner_name") or "Master"


# ---------------------------------------------------------------------------
# Puppy token
# ---------------------------------------------------------------------------


@_registered_cache
def get_puppy_token() -> str | None:
    """Returns the puppy_token from config, or None if not set."""
    return get_value("puppy_token")


def set_puppy_token(token: str) -> None:
    """Sets the puppy_token in the persistent config file."""
    set_config_value("puppy_token", token)


# ---------------------------------------------------------------------------
# Agent directories
# ---------------------------------------------------------------------------


def get_user_agents_directory() -> str:
    """Return the user-level agents directory. Ensures it exists.

    ADR-003: Resolves via _path_agents_dir which respects pup-ex isolation.
    """
    agents_dir = str(_path_agents_dir())
    if not os.path.exists(agents_dir):
        from code_puppy.config_paths import safe_mkdir_p

        safe_mkdir_p(agents_dir)
    return agents_dir


def get_project_agents_directory() -> str | None:
    """Return the project-local agents directory if it exists, or None.

    Project-local agents are NOT subject to the isolation guard (they
    live in ``.code_puppy/agents/`` in the CWD).
    """
    path = os.path.join(os.getcwd(), ".code_puppy", "agents")
    if os.path.isdir(path):
        return path
    return None
